test/sparse/higherorderfns.jl

```diff
@@ -0,0 +1,250 @@
+    for shapeA in ((N,), (N, M))
+        A, B, Co = sprand(shapeA..., 0.2), sprand(shapeA..., 0.2), sprand(shapeA..., 0.2)
+        C = ndims(Co) == 1 ? SparseVector{Float32,Int32}(Co) : SparseMatrixCSC{Float32,Int32}(Co)
+        # use different types to check internal type stability via allocation tests below
+        fA, fB, fC = map(Array, (A, B, C))
+        # --> test map entry point
+        @test map(+, A, B, C) == sparse(map(+, fA, fB, fC))
+        @test map(*, A, B, C) == sparse(map(*, fA, fB, fC))
+        @test map(f, A, B, C) == sparse(map(f, fA, fB, fC))
+        @test_throws DimensionMismatch map(+, A, B, spzeros(N, M - 1))
+        # --> test map! entry point
+        fX = map(+, fA, fB, fC); X = sparse(fX)
+        map!(+, X, A, B, C); X = sparse(fX) # warmup for @allocated
+        @test (@allocated map!(+, X, A, B, C)) == 0
+        @test map!(+, X, A, B, C) == sparse(map!(+, fX, fA, fB, fC))
+        fX = map(*, fA, fB, fC); X = sparse(fX)
+        map!(*, X, A, B, C); X = sparse(fX) # warmup for @allocated
+        @test (@allocated map!(*, X, A, B, C)) == 0
+        @test map!(*, X, A, B, C) == sparse(map!(*, fX, fA, fB, fC))
+        @test map!(f, X, A, B, C) == sparse(map!(f, fX, fA, fB, fC))
+        @test_throws DimensionMismatch map!(f, X, A, B, spzeros((shapeA .- 1)...))
+    end
+end
+
+@testset "broadcast[!] implementation specialized for a single (input) sparse vector/matrix" begin
+    # broadcast[!] for a single sparse vector/matrix falls back to map[!], tested extensively
+    # above. here we simply lightly exercise the relevant broadcast[!] entry points.
+    N, M, p = 10, 12, 0.4
+    a, A = sprand(N, p), sprand(N, M, p)
+    fa, fA = Array(a), Array(A)
+    @test broadcast(sin, a) == sparse(broadcast(sin, fa))
+    @test broadcast(sin, A) == sparse(broadcast(sin, fA))
+    @test broadcast!(sin, copy(a), a) == sparse(broadcast!(sin, copy(fa), fa))
+    @test broadcast!(sin, copy(A), A) == sparse(broadcast!(sin, copy(fA), fA))
+end
+
+@testset "broadcast[!] implementation specialized for pairs of (input) sparse vectors/matrices" begin
+    N, M, p = 10, 12, 0.3
+    f(x, y) = x + y + 1
+    mats = (sprand(N, M, p), sprand(N, 1, p), sprand(1, M, p), sprand(1, 1, 1.0), spzeros(1, 1))
+    vecs = (sprand(N, p), sprand(1, 1.0), spzeros(1))
+    tens = (mats..., vecs...)
+    for Xo in tens
+        X = ndims(Xo) == 1 ? SparseVector{Float32,Int32}(Xo) : SparseMatrixCSC{Float32,Int32}(Xo)
+        # use different types to check internal type stability via allocation tests below
+        shapeX, fX = size(X), Array(X)
+        for Y in tens
+            fY = Array(Y)
+            # --> test broadcast entry point
+            @test broadcast(+, X, Y) == sparse(broadcast(+, fX, fY))
+            @test broadcast(*, X, Y) == sparse(broadcast(*, fX, fY))
+            @test broadcast(f, X, Y) == sparse(broadcast(f, fX, fY))
+            try Base.Broadcast.broadcast_indices(spzeros((shapeX .- 1)...), Y)
```

Comment:
Is there another way of determining shape compatibility than checking for something throwing? Calling some internal boolean function maybe? Otherwise seems like you might not be checking carefully enough which inputs do or do not trigger the test